### DISCUSSION TITLE
atomWithStorage behaviour different in v1 and v2

### DISCUSSION THREAD

**Discussion by adityarupifi** [pmndrs/jotai#1879] [Q&A]
restoring value behaviour in v1
last updated value is read from browser storage on refresh
in this case "new"

<img width="1391" alt="Screenshot 2023-04-12 at 9 15 56 PM" src="https://user-images.githubusercontent.com/107840653/231512665-fcbae4b2-c233-4919-99fb-51c9f930723c.png">

restoring value behaviour in v2
initial value is read from browser storage on refresh then asynchronously updated with last value
in this case "test"

<img width="1391" alt="Screenshot 2023-04-12 at 9 16 45 PM" src="https://user-images.githubusercontent.com/107840653/231512750-1499feb0-525f-4ebf-b58c-9e5128534a45.png">

atomWithStorage usage

<img width="1391" alt="Screenshot 2023-04-12 at 9 17 14 PM" src="https://user-images.githubusercontent.com/107840653/231512798-a5233db4-3fe1-4182-bc46-7781a8fd90dc.png">

Why this behaviour is changed? Is this a bug ?

**Comment by dai-shi** [comment]
This change is intentional, which is called `delayInit` in v1.
The rationale behind it is that it can avoid hydration errors. (Otherwise, there are no workarounds.)

But, for users who don't use SSR, the delayInit behavior doesn't seem ideal.
We suggest to create `atomWithAsyncStorage` and `atomWithWebStorage` instead, and put them in third-party packages.
(Because supporting/maintaining everything in a single function was a disaster in v1.)

AFAIR, someone developed `atomWithAsyncStorage`. We hope to include it in https://github.com/jotaijs/jotai-suspense
For the latter one, I'm not exactly sure which library should have it. Maybe `jotai-persist`?

**Comment by dai-shi** [comment]
> AFAIR, someone developed atomWithAsyncStorage.

https://github.com/pmndrs/jotai/discussions/1784

Would anyone like to open a PR in https://github.com/jotaijs/jotai-suspense ?

**Comment by dai-shi** [comment]
> For the latter one, I'm not exactly sure which library should have it.

Let's have it in persistence guide doc or atom creator recipe doc.

It should be something like this:

```ts
export function atomWithWebStorage(key: string, initialValue: string, storage = localStorage) {
  const baseAtom = atom(storage.getItem(key) ?? initialValue)
  return atom(
    (get) => get(baseAtom),
    (get, set, nextValue: string) => {
      set(baseAtom, nextValue)
      storage.setItem(key, nextValue)
    }
  )
}
```

Would anyone like to try?

**Reply by arjunvegda** [reply]
Since these are storage related, should we put both (Async and Web) in `jotai-persist`? 🤔  

Also, I ran into this today, this is what I came up with https://codesandbox.io/s/atom-with-web-storage-poc-itrulp?file=/src/atomWithWebStorage.ts

**Reply by dai-shi** [reply]
On second thought, `jotai-persist` sounds too official and the concern is people get confused which one to use.

My assumption is, `atomWithWebStorage` isn't rarely necessary and `atomWithStorage` works for _most_ web cases, while `atomWithAsyncStorage` is more desirable for _most_ RN cases than `atomWithStorage`.
I didn't realize this issue in RN when preparing v2.
If I had known it, we could have deprecated AsyncStorage support in `atomWithStorage` and provide the separate `atomWithAsyncStorage`.

So, if we go with `jotai-persist`, I want `atomWithStorage` in `jotai` to be deprecated.
